hpcrawler.py

```diff
@@ -0,0 +1,364 @@
+                the track's representation:  dict {<country>_<j>: [slopes, track_dif]}
+        """
+        # print("\t* processing track") #TODO
+        points, track_elev = pd.DataFrame({}), pd.DataFrame({})
+        filename, track_len, track_dif = features
+        track_len = float(track_len)
+
+        # open gpx file:
+        gpx_file = open(os.path.join(HpCrawler.gpx_dir_path, self._country, self._track_idx + ".gpx"),
+                        'r', encoding="utf8")
+        gpx = gpxpy.parse(gpx_file)
+
+        # get points array & elevations:
+        for track in gpx.tracks:
+            for seg in track.segments:
+                points = pd.DataFrame([{'lat': p.latitude, 'lon': p.longitude} for p in seg.points]) \
+                    .to_numpy()
+                track_elev = pd.DataFrame([{'ele': p.elevation} for p in seg.points]) \
+                    .to_numpy().reshape(len(points))
+
+        # compute slopes:
+        tick = sm.get_tick(track_len)
+        slopes = sm.compute_slope(points, track_elev, tick)
+
+        # save track to the correct dict (according to the track's length):
+        len_tag = sm.get_length_tag(track_len)
+
+        return len_tag, {self._country + '_' + self._track_idx: [slopes, track_dif]}
+
+# runs functionality:
+
+    def crawl(self):
+        """
+        crawls the trails data from "the hiking project"'s site.
+        it saves the data and the progress of crawling under the "hp" directory.
+        :return all of the countries we've crawled so far (not just in this iteration)
+        """
+        for i in range(len(self._countries)):
+            seen = HpCrawler.load_seen()
+            self._country = self._countries[i]
+
+            if self._country in seen and seen[self._country] == HpCrawler.done_tag:  # country done processing
+                continue
+
+            print("\n", self._country)
+
+            # create gpx folder for country of index i, sets up driver to website:
+            self._driver = self._setup()
+
+            # (try and try until you) log in
+            while 1:
+                try:
+                    self._driver.implicitly_wait(HpCrawler.wait)
+                    self._log_in()
+                    break
+                except selenium.common.exceptions.ElementNotInteractableException:  # race condition- driver not ready
+                    print("USAGE: Do not minimize the window until logged in.")
+
+            if self._country in seen:  # country mid-processing:
+                self._track_idx, trail_urls = seen[self._country]
+            else:  # new country!
+                trail_urls = self._trails_in_urls()
+                seen.update({self._country: [self._track_idx, trail_urls]})
+                HpCrawler._save_dict(seen, HpCrawler.seen_path)
+
+            # start mining:
+            print("-- collecting and processing")
+
+            for j in np.arange(int(self._track_idx), (len(trail_urls))):  # trails we didn't finish processing
```

Comment:
I'd personally like this loop in its own function since this function is quite long and a bit messy